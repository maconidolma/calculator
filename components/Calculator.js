import React, {Component} from 'react'
import Screen from './Screen'
import Keyboard from './Keyboard'

import fetch from 'isomorphic-fetch'



class Calculator extends Component {
    constructor(props) {
        super(props);

        this.buttonHandler = this.buttonHandler.bind(this);

        this.state = {
            currentScreenValue: 0,
            firstOperand: '',
            secondOperand: '',
            operation: null,
            isFirstOperand: true,
            shouldClrScreen: true,
            prevEqually: false
        }
    }

    render() {
        return (
            <div className={'calculator'}>
                <div>
                    <Screen currentScreenValue = {this.state.currentScreenValue}/>
                </div>
                <div>
                    <Keyboard buttonHandler = {this.buttonHandler}/>
                </div>
            </div>
        )
    }

    sendToServer(str) {
        console.log(str);

        let url = 'http://localhost:3000/expressions';
        let data = {operation: str};

        fetch(url, {
            method: 'POST',
            body: JSON.stringify(data),
            headers: {
                'Content-Type': 'application/json'
            }
        }).then(response => console.log('Success: ', response))
            .catch(function(error) {
                 //console.log(error);
            })
    }

    calculateAndSendToServer(fOp, sOp, op ) {
        console.log('hello');
        let calculatedValue = null;
        switch (op) {
            case '+' :
                calculatedValue = fOp + sOp;
                break;
            case '-' :
                calculatedValue = fOp - sOp;
                break;
            case '*' :
                calculatedValue = fOp * sOp;
                break;
            case '/' :
                calculatedValue = fOp / sOp;
                break;
        }

        this.sendToServer(fOp.toString() + ' ' + op + ' ' + sOp.toString() +  ' = ' + calculatedValue.toString());
        return calculatedValue;
    }  // calculateAndSendToServer()

    buttonHandler(buttonValue) {
        let newState = {};

        if (this.state.shouldClrScreen) {
            newState = {
                ...newState,
                currentScreenValue: '',
                shouldClrScreen: false
            }
        }

        if (!isNaN(buttonValue)) {  // If value is number

            // If prev buttonValue was '=' and current buttonValue is number
            if (this.state.prevEqually) {
                newState = {
                    ...newState,
                    firstOperand: buttonValue,
                    isFirstOperand: true,
                    currentScreenValue: buttonValue,
                    prevEqually: false
                }
            } else if (this.state.isFirstOperand) {  // Add number to screen and proper operand
                newState = {
                    ...newState,
                    firstOperand: this.state.firstOperand + buttonValue,
                    currentScreenValue: this.state.firstOperand + buttonValue
                }
            } else {
                newState = {
                    ...newState,
                    secondOperand: this.state.secondOperand + buttonValue,
                    currentScreenValue: this.state.secondOperand + buttonValue

                }
            }

        } else if (buttonValue === '+'
                   || buttonValue === '-'
                   || buttonValue === '*'
                   || buttonValue === '/') {  // If value is operation

            if (this.state.isFirstOperand) {
                newState = {
                    ...newState,
                    operation: buttonValue,
                    isFirstOperand: false,
                    prevEqually: false
                }
            } else {
                let res = this.calculateAndSendToServer(parseInt(this.state.firstOperand),
                    parseInt(this.state.secondOperand),
                    this.state.operation);
                newState = {
                    ...newState,
                    currentScreenValue: res,
                    firstOperand: res,
                    secondOperand: '',
                    isFirstOperand: false,
                    shouldClrScreen: true,
                    operation: buttonValue
                }

            }
        } else if (buttonValue === '=') {
            let res1 = this.calculateAndSendToServer(parseInt(this.state.firstOperand),
                                                     parseInt(this.state.secondOperand),
                                                     this.state.operation);
            newState = {
                ...newState,
                currentScreenValue: res1,
                firstOperand: res1,
                secondOperand: '',
                isFirstOperand: true,
                operation: null,
                prevEqually: true
            }

        } else if (buttonValue === 'AC') {
            newState = {
                ...newState,
                currentScreenValue: 0,
                firstOperand: '',
                secondOperand: '',
                operation: null,
                isFirstOperand: true
            }
        }

        this.setState ({
            ...newState
        });

        console.log('currentState', newState);
    }  // buttonHandler()
}

export default Calculator