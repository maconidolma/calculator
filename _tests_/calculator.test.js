import React from 'react';
import { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';


import  Calculator from '../components/Calculator';





function mouneAndFind() {
    const wrapper = mount(
        <Calculator  />
    );
    let buttons = {};
    buttons['AC'] = wrapper.find('.boxAC');
    buttons['/'] = wrapper.find('.boxDiv');
    buttons['7'] = wrapper.find('.box7');
    buttons['8'] = wrapper.find('.box8');
    buttons['9'] = wrapper.find('.box9');
    buttons['4'] = wrapper.find('.box4');
    buttons['5'] = wrapper.find('.box5');
    buttons['6'] = wrapper.find('.box6');
    buttons['-'] = wrapper.find('.boxMn');
    buttons['1'] = wrapper.find('.box1');
    buttons['2'] = wrapper.find('.box2');
    buttons['3'] = wrapper.find('.box3');
    buttons['+'] = wrapper.find('.boxPl');
    buttons['0'] = wrapper.find('.box0');
    buttons['='] = wrapper.find('.boxEq');
    buttons['*'] = wrapper.find('.boxMul');
    buttons['Cur'] = wrapper.find('.boxCur');
    const screen = wrapper.find('.screen');
    return [buttons, screen, wrapper];
}


test('CalculatorComponent renders the numbers at screen', () => {
    let [buttons, screen] = mouneAndFind();

    buttons['5'].simulate('click');
    expect(screen.text()).toBe('5');

    buttons['-'].simulate('click');

    buttons['5'].simulate('click');
    buttons['5'].simulate('click');
    expect(screen.text()).toBe('55');

    buttons['*'].simulate('click');
    expect(screen.text()).toBe('-50');

    buttons['5'].simulate('click');
    expect(screen.text()).toBe('5');

    buttons['='].simulate('click');
    expect(screen.text()).toBe('-250');
});


test('CalculatorComponent', () => {
    let [buttons, screen, wrapper] = mouneAndFind();

    expect(shallowToJson(wrapper)).toMatchSnapshot();
});

test('CalculatorState', () => {
    let [buttons, screen, wrapper] = mouneAndFind();

    expect(wrapper.state().firstOperand).toEqual('');

    buttons['5'].simulate('click');
    expect(wrapper.state().firstOperand).toEqual('5');
});


