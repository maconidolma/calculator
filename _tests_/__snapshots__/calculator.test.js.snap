// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorComponent 1`] = `
<Calculator>
  <div
    className="calculator"
  >
    <div>
      <Screen
        currentScreenValue={0}
      >
        <div
          className="screen"
        >
          0
        </div>
      </Screen>
    </div>
    <div>
      <Keyboard
        buttonHandler={[Function]}
      >
        <div
          className="keyBoard"
        >
          <Button
            buttonHandler={[Function]}
            buttonValue="AC"
            clName="box boxAC"
          >
            <div
              className="box boxAC"
              onClick={[Function]}
            >
              AC
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="/"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              /
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="7"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              7
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="8"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              8
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="9"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              9
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="*"
            clName="box boxMul"
          >
            <div
              className="box boxMul"
              onClick={[Function]}
            >
              *
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="4"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              4
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="5"
            clName="box box5"
          >
            <div
              className="box box5"
              onClick={[Function]}
            >
              5
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="6"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              6
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="-"
            clName="box boxMn"
          >
            <div
              className="box boxMn"
              onClick={[Function]}
            >
              -
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="1"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              1
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="2"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              2
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="3"
            clName="box"
          >
            <div
              className="box"
              onClick={[Function]}
            >
              3
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="+"
            clName="box boxPl"
          >
            <div
              className="box boxPl"
              onClick={[Function]}
            >
              +
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="0"
            clName="box box0"
          >
            <div
              className="box box0"
              onClick={[Function]}
            >
              0
            </div>
          </Button>
          <Button
            buttonHandler={[Function]}
            buttonValue="="
            clName="box boxEq"
          >
            <div
              className="box boxEq"
              onClick={[Function]}
            >
              =
            </div>
          </Button>
        </div>
      </Keyboard>
    </div>
  </div>
</Calculator>
`;
